pyflakes/test/test_other.py

```diff
@@ -13,6 +13,97 @@ class Test(TestCase):
     def test_duplicateArgs(self):
         self.flakes('def fu(bar, bar): pass', m.DuplicateArgument)
 
+    def test_duplicate_keys(self):
+        self.flakes("{'yes': 1, 'yes': 1}", m.DuplicateDictionaryKey)
+
+    def test_multiple_duplicate_keys(self):
+        self.flakes(
+            "{'yes': 1, 'yes': 1, 'no': 2, 'no': 3}",
+            m.DuplicateDictionaryKey, m.DuplicateDictionaryKey)
+
+    def test_duplicate_keys_in_function(self):
+        self.flakes('''
+        def f(thing):
+            pass
+        f({'yes': 1, 'yes': 4})
+        ''', m.DuplicateDictionaryKey)
+
+    def test_duplicate_keys_in_lambda(self):
+        self.flakes("lambda x: {(0,1): 1, (0,1): 1}",
+                    m.DuplicateDictionaryKey)
+
+    def test_duplicate_keys_tuples(self):
+        self.flakes("{(0,1): 1, (0,1): 1}", m.DuplicateDictionaryKey)
+
+    def test_duplicate_keys_ints(self):
+        self.flakes("{1: 1, 1: 1}", m.DuplicateDictionaryKey)
+
+    def test_duplicate_keys_bools(self):
+        if version_info < (3, 4):
+            expected = m.DuplicateVariableDictionaryKey
+        else:
+            expected = m.DuplicateDictionaryKey
+        self.flakes("{True: 1, True: 1}", expected)
+
+    def test_duplicate_keys_none(self):
+        if version_info < (3, 4):
+            expected = m.DuplicateVariableDictionaryKey
+        else:
+            expected = m.DuplicateDictionaryKey
+        self.flakes("{None: 1, None: 1}", expected)
+
+    def test_duplicate_variable_keys(self):
+        self.flakes('''
+        a = 1
+        {a: 2, a: 3}
+        ''', m.DuplicateVariableDictionaryKey)
+
+    def test_duplicate_key_float_and_int(self):
```

Comment:
This one really surprised me, as they are stored differently
```
>>> {1.0: 'a'}
{1.0: 'a'}
>>> {1: 'a'}
{1: 'a'}
```
The same occurs for sets.
```
>>> {1, 1.0}
set([1.0])
```

But it is documented behaviour (https://github.com/python/cpython/blame/master/Doc/library/stdtypes.rst#L4008) . (Probably would be useful for the Python reference on dict/set displays to mention it also)

Anyway, it will be very helpful for pyflakes to be catching this oddity.  IMO it justifies any 'false positives' that this improvement might have.

IMO this test deserves a docstring ;-)